### ISSUE TITLE
exhaustive match checking on nested class pattern

### ISSUE THREAD

**Issue by pinterior** [microsoft/pyright#10894]
**Is your feature request related to a problem? Please describe.**

Currently, when using `match`-`case` is on a nested data structure, `reportMatchNotExhaustive` is reported even if the match is exhaustive.

```py
# pyright: reportMatchNotExhaustive=true

from dataclasses import dataclass
from typing import final

@final
@dataclass
class Some[T]:
    value: T

@final
@dataclass
class Nothing:
    pass

type Maybe[T] = Some[T] | Nothing

def f(x: Maybe[Maybe[int]]):
    match x:
        case Some(Some(_)) | Some(Nothing()) | Nothing():
            pass
```

```
$ pyright --version
pyright 1.1.405
$ pyright bar.py 
bar.py
  bar.py:19:11 - error: Cases within match statement do not exhaustively handle all values
    Unhandled type: "Some[Maybe[int]]"
    If exhaustive handling is not intended, add "case _: pass" (reportMatchNotExhaustive)
1 error, 0 warnings, 0 informations
```

**Describe the solution you’d like**

It would be great if the exhaustive check worked like in languages such as Scala.

**Comment by erictraut** [comment]
This behavior is expected. Exhaustion detection is based on type narrowing, and the type of `x` cannot be narrowed to `Never` in this case.

If you want to benefit from exhaustion detection in a type checker, you'll need to avoid this pattern.
